src/core/ext/filters/client_channel/resolver/binder/binder_resolver.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/string_util.h>
+
+#include "src/core/ext/filters/client_channel/resolver_registry.h"
+#include "src/core/ext/filters/client_channel/server_address.h"
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gpr/string.h"
+#include "src/core/lib/iomgr/resolve_address.h"
+#include "src/core/lib/iomgr/unix_sockets_posix.h"
+
+namespace grpc_core {
+namespace {
+
+class BinderResolver : public Resolver {
+ public:
+  BinderResolver(ServerAddressList addresses, ResolverArgs args)
+      : result_handler_(std::move(args.result_handler)),
+        addresses_(std::move(addresses)),
+        channel_args_(grpc_channel_args_copy(args.args)) {}
+
+  ~BinderResolver() override { grpc_channel_args_destroy(channel_args_); };
+
+  void StartLocked() override {
+    Result result;
+    result.addresses = std::move(addresses_);
+    result.args = channel_args_;
+    channel_args_ = nullptr;
+    result_handler_->ReturnResult(std::move(result));
+  }
+
+  void ShutdownLocked() override {}
+
+ private:
+  std::unique_ptr<ResultHandler> result_handler_;
+  ServerAddressList addresses_;
+  const grpc_channel_args* channel_args_ = nullptr;
+};
+
+grpc_error_handle BinderAddrPopulate(const std::string path,
+                                     grpc_resolved_address* resolved_addr) {
+  if (path.empty()) {
+    return GRPC_ERROR_CREATE_FROM_CPP_STRING("path is empty");
+  }
+  // See README for valid characters
+  if (std::find_if(path.begin(), path.end(), [](char c) {
+        return !(isalnum(c) || absl::StrContains("-._~", c));
+      }) != path.end()) {
+    return GRPC_ERROR_CREATE_FROM_CPP_STRING(path +
+                                             " contains invalid character");
+  }
+  if (path.size() + 1 > sizeof(resolved_addr->addr)) {
+    return GRPC_ERROR_CREATE_FROM_CPP_STRING(path +
+                                             " is too long to be handled");
+  }
+  strcpy(resolved_addr->addr, path.c_str());
```

Comment:
There's a lot of code in gRPC that sees the address after it is returned from the resolver and before it gets passed to `BinderClientChannelFactory::CreateSubchannel()` to create the connection.  All of that code treats this address as a sockaddr and will try to interact with it as such.  If you just populate it as a string, how will any of that code react?

@ctiller I think that in order to support this, we may need to build some of the infrastructure you proposed for different address types.  This may need more discussion.

Reply:
Yeah I saw there are some code that attempts to parse `grpc_resolved_address` to string for logging purposes (which is probably acceptable for us). However if there are code that tries to reinterpret `grpc_resolved_address` as `sockaddr` and tries to use that address to actually do something it would be pretty bad.

This does not sound like an issue that can be resolved quickly. How could we proceed here? How about saving our "binder identifier string" in a `sockaddr` struct and make sure it have a invalid address family (e.g `AF_MAX`) so there is no security concern about accidentally connecting to some random socket?